katharsis-jpa/src/main/java/io/katharsis/jpa/internal/meta/impl/MetaDataObjectImpl.java

```diff
@@ -0,0 +1,376 @@
+		return attrs;
+	}
+
+	@Override
+	public List<MetaAttribute> getDeclaredAttributes() {
+		return declaredAttrs;
+	}
+
+	@Override
+	public MetaAttribute getAttribute(String name) {
+		MetaAttributeImpl attr = attrMap.get(name);
+		KatharsisAssert.assertNotNull(getName() + "." + name, attr);
+		return attr;
+	}
+
+	@Override
+	public String toString(Object entity) {
+		boolean notFirst = false;
+		StringBuilder b = new StringBuilder(entity.getClass().getSimpleName());
+		b.append('[');
+
+		for (MetaAttribute attr : getAttributes()) {
+			if (attr.isAssociation()) {
+				continue;
+			}
+
+			Object value = attr.getValue(entity);
+
+			if (notFirst) {
+				b.append(',');
+			} else {
+				notFirst = true;
+			}
+			b.append(attr.getName());
+			b.append("=");
+			b.append(formatValue(value));
+		}
+		b.append(']');
+		return b.toString();
+	}
+
+	private Object formatValue(Object value) {
+		if (value == null) {
+			return "null";
+		} else if (value instanceof Calendar) {
+			Calendar cal = (Calendar) value;
+			SimpleDateFormat sdf = new SimpleDateFormat("yyyy.MM.dd HH:mm:ss Z");
+			return sdf.format(cal.getTime());
+		} else if (value instanceof Date) {
+			Date cal = (Date) value;
+			SimpleDateFormat sdf = new SimpleDateFormat("yyyy.MM.dd HH:mm:ss Z");
+			return sdf.format(cal.getTime());
+		} else {
+			return value.toString();
+		}
+	}
+
+	protected MetaAttributeImpl newAttributeAttribute(MetaDataObjectImpl metaDataObject, PropertyDescriptor desc) {
+		return new MetaAttributeImpl(metaDataObject, desc);
+	}
+
+	@Override
+	public MetaAttributePath resolvePath(List<String> attrPath, boolean includeSubTypes) {
+		MetaAttributeFinder finder = includeSubTypes ? SUBTYPE_ATTRIBUTE_FINDER : DEFAULT_ATTRIBUTE_FINDER;
+		return resolvePath(attrPath, finder);
+	}
+
+	@Override
+	public MetaAttributePath resolvePath(List<String> attrPath) {
+		return resolvePath(attrPath, true);
+	}
+
+	@Override
+	public MetaAttributePath resolvePath(List<String> attrPath, MetaAttributeFinder finder) {
+		if (attrPath == null || attrPath.isEmpty())
+			throw new IllegalArgumentException("invalid attribute path '" + attrPath + "'");
+		LinkedList<MetaAttribute> list = new LinkedList<>();
+
+		MetaDataObject currentMdo = this;
+		for (int i = 0; i < attrPath.size(); i++) {
+			String pathElementName = attrPath.get(i);
+			MetaAttribute pathElement = finder.getAttribute(currentMdo, pathElementName);
+			if (i < attrPath.size() - 1 && pathElement.getType() instanceof MetaMapType) {
+				MetaMapType mapType = (MetaMapType) pathElement.getType();
+
+				// next "attribute" is the key within the map
+				String keyString = attrPath.get(i + 1);
+
+				MetaMapAttribute keyAttr;
+				if (pathElement instanceof MetaAttributeProjection) {
+					keyAttr = new MetaMapAttributeProjectionImpl(mapType, pathElement, keyString, true);
+				} else {
+					keyAttr = new MetaMapAttributeImpl(mapType, pathElement, keyString, true);
+				}
+				list.add(keyAttr);
+				i++;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor the code in order to not assign to this loop counter from within the loop body. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AForLoopCounterChangedCheck)